src/orchestrator.ts

```diff
@@ -200,40 +231,214 @@ export class Orchestrator implements Callbacks {
+        sendingStream: AsyncIterable<StreamChunk> &
+            grpc.ServerDuplexStream<StreamChunk, ReceivingStreamControl>,
+    ): Promise<number> {
+        const globalSequenceNumber = this.globalSequenceNumber++
+        const writeToSender = promisify(sendingStream.write.bind(sendingStream))
+
+        const sourceRunner = this.instantiators[streamIdentify.runner]
+        if (!sourceRunner) {
+            throw (
+                'Failed to find correct source runner with uri ' +
+                streamIdentify.runner
+            )
+        }
+
+        const targetInstantiator =
+            this.channelToInstantiator[streamIdentify.channel]
+
+        // If the instantiator exists, it will send a processed message
+        // When this happens, the onMessageProcessedCb will be called
+        // If the instantiator does not exist, it cannot send a processed message
+        // So the onMessageProcessedCb should be called when the stream is closed on the writer's side
+        if (targetInstantiator) {
+            this.runningMessages[globalSequenceNumber] =
+                sourceRunner.onMessageProcessedCb(
+                    streamIdentify.localSequenceNumber,
+                    streamIdentify.channel,
+                )
+
+            const readerConnected = new Promise<ReceivingStream>((res) => {
+                this.waitForConnectingReceivingStream[globalSequenceNumber] =
+                    res
+            })
+
+            targetInstantiator.streamMessage({
+                channel: streamIdentify.channel,
+                globalSequenceNumber,
+            })
+
+            const receivingStream = await readerConnected
+            const writeToReceiver = promisify(
+                receivingStream.write.bind(receivingStream),
+            )
+            const log = this.logLevels[streamIdentify.channel]
+
+            sendingStream.on('data', async (chunk) => {
+                if (chunk.data !== undefined) {
+                    if (log) {
+                        log(decoder.decode(chunk.data!.data))
+                    }
+                    await writeToReceiver(chunk.data)
+                }
+            })
+
+            sendingStream.on('end', () => {
+                try {
+                    receivingStream.end()
+                } catch (ex) {
+                    if (ex instanceof Error) {
+                        console.error(ex.stack)
+                    }
+                }
+            })
+
+            receivingStream.on('data', async (d) => {
+                try {
+                    await writeToSender(d)
+                } catch (ex) {
+                    console.log('Error happend', ex)
+                }
+            })
+
+            writeToSender({ streamSequenceNumber: globalSequenceNumber })
+        } else {
+            this.logger.error(
+                `Receiving stream message for channel ${streamIdentify.channel} without a connected reader`,
```

Comment:
Is this (and the normal message variant) actually an error, or should it rather be a warning?

Reply:
But if we consider e.g. the shacl-processor which outputs both the validated stream of data and the validation report as output. Imagine I am just interested in the validated data, and don't care about the validation report, would you call it an invalid pipeline then if I don't consume the validation report I'm not interested in?

Reply:
That's also an option indeed 👍